File: Uas/Caffèra — Cafe Management System/Backend/Services/Implementations/OrderService.cs
using Caffera.Backend.DTOs.Common;
using Caffera.Backend.DTOs.Order;
using Caffera.Backend.Models;
using Caffera.Backend.Repositories.Interfaces;
using Caffera.Backend.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Caffera.Backend.Services.Implementations;

public class OrderService : IOrderService
{
    private readonly IOrderRepository _orderRepository;
    private readonly IRepository<Menu> _menuRepository;
    private readonly IRepository<Table> _tableRepository;
    private readonly IRepository<User> _userRepository;
    private readonly ILogger<OrderService> _logger;

    public OrderService(
        IOrderRepository orderRepository,
        IRepository<Menu> menuRepository,
        IRepository<Table> tableRepository,
        IRepository<User> userRepository,
        ILogger<OrderService> logger)
    {
        _orderRepository = orderRepository;
        _menuRepository = menuRepository;
        _tableRepository = tableRepository;
        _userRepository = userRepository;
        _logger = logger;
    }

    public async Task<ApiResponse<PagedResult<OrderDto>>> GetAllAsync(OrderFilterParams filterParams)
    {
        _logger.LogInformation("[ORDER] 📋 Mengambil daftar pesanan (Page: {Page}, Size: {Size}, Status: '{Status}', Type: '{Type}', Search: '{Search}')",
            filterParams.PageNumber, filterParams.PageSize, filterParams.Status ?? "All", filterParams.OrderType ?? "All", filterParams.Search ?? "-");

        var query = _orderRepository.Query()
            .Include(o => o.User)
            .Include(o => o.Table)
            .Include(o => o.OrderItems)
                .ThenInclude(oi => oi.Menu!)
                    .ThenInclude(m => m.Category)
            .AsQueryable();

        // 1. Search filter (OrderNumber, Customer/User, Table)
        if (!string.IsNullOrWhiteSpace(filterParams.Search))
        {
            var search = filterParams.Search.Trim().ToLower();
            query = query.Where(o => o.OrderNumber.ToLower().Contains(search) || 
                                     (o.User != null && o.User.Name.ToLower().Contains(search)) ||
                                     (o.CustomerName != null && o.CustomerName.ToLower().Contains(search)));
        }

        // 2. Status filter
        if (!string.IsNullOrWhiteSpace(filterParams.Status))
        {
            var rawStatus = filterParams.Status.Trim();
            if (!rawStatus.Equals("all", StringComparison.OrdinalIgnoreCase))
            {
                if (rawStatus.Equals("active", StringComparison.OrdinalIgnoreCase) ||
                    rawStatus.Equals("kitchen", StringComparison.OrdinalIgnoreCase) ||
                    rawStatus.Equals("inprogress", StringComparison.OrdinalIgnoreCase) ||
                    rawStatus.Equals("in_progress", StringComparison.OrdinalIgnoreCase))
                {
                    query = query.Where(o => o.Status == "Pending" || o.Status == "Preparing" || o.Status == "Ready");
                }
                else
                {
                    var statuses = rawStatus.Split(',', StringSplitOptions.RemoveEmptyEntries | StringSplitOptions.TrimEntries)
                                            .Select(s => s.ToLower())
                                            .ToList();
                    if (statuses.Count == 1)
                    {
                        var singleStatus = statuses[0];
                        query = query.Where(o => o.Status.ToLower() == singleStatus);
                    }
                    else if (statuses.Count > 1)
                    {
                        query = query.Where(o => statuses.Contains(o.Status.ToLower()));
                    }
                }
            }
        }

        // 3. OrderType filter
        if (!string.IsNullOrWhiteSpace(filterParams.OrderType) && !filterParams.OrderType.Equals("all", StringComparison.OrdinalIgnoreCase))
        {
            query = query.Where(o => o.OrderType.ToLower() == filterParams.OrderType.Trim().ToLower());
        }

        // 4. Date range filter
        if (filterParams.StartDate.HasValue)
        {
            query = query.Where(o => o.CreatedAt >= filterParams.StartDate.Value.Date);
        }

        if (filterParams.EndDate.HasValue)
        {
            var endOfDay = filterParams.EndDate.Value.Date.AddDays(1).AddTicks(-1);
            query = query.Where(o => o.CreatedAt <= endOfDay);
        }

        var totalCount = await query.CountAsync();

        // 5. Sorting
        query = filterParams.SortBy?.ToLower() switch
        {
            "totalamount" => filterParams.IsAscending ? query.OrderBy(o => o.TotalAmount) : query.OrderByDescending(o => o.TotalAmount),
            "ordernumber" => filterParams.IsAscending ? query.OrderBy(o => o.OrderNumber) : query.OrderByDescending(o => o.OrderNumber),
            "status" => filterParams.IsAscending ? query.OrderBy(o => o.Status) : query.OrderByDescending(o => o.Status),
            _ => query.OrderByDescending(o => o.CreatedAt)
        };

        // 6. Pagination
        var skip = (filterParams.PageNumber - 1) * filterParams.PageSize;
        var orders = await query.Skip(skip).Take(filterParams.PageSize).ToListAsync();

        var dtos = orders.Select(MapToOrderDto);

        var pagedResult = new PagedResult<OrderDto>
        {
            Items = dtos,
            PageNumber = filterParams.PageNumber,
            PageSize = filterParams.PageSize,
            TotalCount = totalCount
        };

        return ApiResponse<PagedResult<OrderDto>>.SuccessResult(pagedResult);
    }

    public async Task<ApiResponse<OrderDto>> GetByIdAsync(int id)
    {
        _logger.LogInformation("[ORDER] 🔍 Mengambil detail pesanan ID: {OrderId}", id);
        var order = await _orderRepository.GetOrderWithDetailsAsync(id);
        if (order == null)
        {
            _logger.LogWarning("[ORDER] ⚠️ Pesanan dengan ID {OrderId} tidak ditemukan", id);
            return ApiResponse<OrderDto>.FailResult("Pesanan tidak ditemukan");
        }

        return ApiResponse<OrderDto>.SuccessResult(MapToOrderDto(order));
    }

    public async Task<ApiResponse<OrderDto>> GetByOrderNumberAsync(string orderNumber)
    {
        _logger.LogInformation("[ORDER] 🔍 Mengambil detail pesanan berdasarkan Nomor: '{OrderNumber}'", orderNumber);
        var order = await _orderRepository.GetOrderByNumberWithDetailsAsync(orderNumber.Trim());
        if (order == null)
        {
            _logger.LogWarning("[ORDER] ⚠️ Pesanan dengan Nomor '{OrderNumber}' tidak ditemukan", orderNumber);
            return ApiResponse<OrderDto>.FailResult("Pesanan tidak ditemukan");
        }

        return ApiResponse<OrderDto>.SuccessResult(MapToOrderDto(order));
    }

    public async Task<ApiResponse<OrderDto>> CreateAsync(int userId, CreateOrderDto dto)
    {
        var normalizedItems = dto.GetNormalizedItems();
        _logger.LogInformation("[ORDER] 🛒 Membuat pesanan baru oleh User ID: {UserId}, Tipe: {OrderType}, Jumlah Menu: {Count}", 
            userId, dto.OrderType, normalizedItems.Count);

        if (normalizedItems.Count == 0)
        {
            return ApiResponse<OrderDto>.FailResult("Pesanan harus memiliki minimal 1 item menu yang valid");
        }

        var user = await _userRepository.GetByIdAsync(userId);
        if (user == null)
        {
            _logger.LogWarning("[ORDER] ⚠️ Gagal membuat pesanan: User ID {UserId} tidak ditemukan di database", userId);
            return ApiResponse<OrderDto>.FailResult("Kasir/Pengguna tidak valid");
        }

        Table? table = null;
        if (dto.OrderType.Equals("DineIn", StringComparison.OrdinalIgnoreCase))
        {
            if (dto.TableId.HasValue && dto.TableId.Value > 0)
            {
                table = await _tableRepository.GetByIdAsync(dto.TableId.Value);
            }
            if (table == null && dto.TableNumber.HasValue && dto.TableNumber.Value > 0)
            {
                var tables = await _tableRepository.FindAsync(t => t.Number == dto.TableNumber.Value);
                table = tables.FirstOrDefault();
            }
            if (table == null && dto.TableId.HasValue && dto.TableId.Value > 0)
            {
                var tables = await _tableRepository.FindAsync(t => t.Number == dto.TableId.Value);
                table = tables.FirstOrDefault();
            }
            if (table == null && dto.TableNumber.HasValue && dto.TableNumber.Value > 0)
            {
                table = await _tableRepository.GetByIdAsync(dto.TableNumber.Value);
            }

            if (table != null)
            {
                dto.TableId = table.Id;
                table.Status = "Occupied";
                table.UpdatedAt = DateTime.UtcNow;
                await _tableRepository.UpdateAsync(table);
                _logger.LogInformation("[ORDER] 🪑 Status Meja #{TableNumber} diubah menjadi Occupied", table.Number);
            }
        }

        var menuIds = normalizedItems.Select(i => i.EffectiveMenuId).Distinct().ToList();
        var menus = (await _menuRepository.FindAsync(m => menuIds.Contains(m.Id))).ToDictionary(m => m.Id);

        foreach (var item in normalizedItems)
        {
            if (!menus.TryGetValue(item.EffectiveMenuId, out var menu))
            {
                _logger.LogWarning("[ORDER] ⚠️ Gagal membuat pesanan: Menu ID {MenuId} tidak ditemukan", item.EffectiveMenuId);
                return ApiResponse<OrderDto>.FailResult($"Menu dengan ID {item.EffectiveMenuId} tidak ditemukan");
            }

            if (!menu.IsAvailable)
            {
                _logger.LogWarning("[ORDER] ⚠️ Gagal membuat pesanan: Menu '{MenuName}' (ID: {MenuId}) sedang Tidak Tersedia", menu.Name, menu.Id);
                return ApiResponse<OrderDto>.FailResult($"Menu '{menu.Name}' saat ini sedang tidak tersedia");
            }
        }

        var orderNumber = await _orderRepository.GenerateOrderNumberAsync();
        var customerName = !string.IsNullOrWhiteSpace(dto.CustomerName)
            ? dto.CustomerName.Trim()
            : (table != null ? $"Tamu Meja #{table.Number}" : user.Name);

        var order = new Order
        {
            OrderNumber = orderNumber,
            UserId = userId,
            CustomerName = customerName,
            TableId = dto.OrderType.Equals("DineIn", StringComparison.OrdinalIgnoreCase) ? (table?.Id ?? dto.TableId) : null,
            OrderType = dto.OrderType.Equals("TakeAway", StringComparison.OrdinalIgnoreCase) ? "TakeAway" : "DineIn",
            Status = "Pending",
            CreatedAt = DateTime.UtcNow,
            OrderItems = new List<OrderItem>()
        };

        decimal totalAmount = 0;
        foreach (var item in normalizedItems)
        {
            var menu = menus[item.EffectiveMenuId];
            var qty = item.EffectiveQuantity;
            var subtotal = menu.Price * qty;
            totalAmount += subtotal;

            order.OrderItems.Add(new OrderItem
            {
                MenuId = menu.Id,
                Quantity = qty,
                Price = menu.Price, // Snapshot price!
                Subtotal = subtotal
            });
        }

        order.TotalAmount = totalAmount;

        var createdOrder = await _orderRepository.AddAsync(order);
        var fullOrder = await _orderRepository.GetOrderWithDetailsAsync(createdOrder.Id);

        _logger.LogInformation("[ORDER] ✅ Pesanan berhasil dibuat! No: {OrderNumber} (ID: {OrderId}), Total: Rp {Total:N0}, Kasir: {Cashier}",
            order.OrderNumber, createdOrder.Id, order.TotalAmount, user.Name);

        var orderDto = MapToOrderDto(fullOrder!);
        return ApiResponse<OrderDto>.SuccessResult(orderDto, "Pesanan berhasil dibuat");
    }

    public async Task<ApiResponse<OrderDto>> CreateGuestOrderAsync(CreateOrderDto dto)
    {
        var normalizedItems = dto.GetNormalizedItems();
        _logger.LogInformation("[ORDER] 📱 Tamu membuat pesanan Self-Order via QR Meja #{TableNumber} / TableId {TableId}. Customer: '{Customer}', Items Count: {Count}", 
            dto.TableNumber, dto.TableId, dto.CustomerName ?? "Tamu", normalizedItems.Count);

        if (normalizedItems.Count == 0)
        {
            return ApiResponse<OrderDto>.FailResult("Pesanan harus memiliki minimal 1 item menu yang valid");
        }

        // Cari akun sistem / admin default sebagai penanggung jawab order
        var users = await _userRepository.GetAllAsync();
        var defaultUser = users.FirstOrDefault(u => u.Role == "Admin") ?? users.FirstOrDefault();

        if (defaultUser == null)
        {
            _logger.LogWarning("[ORDER] ⚠️ Gagal membuat pesanan guest: Tidak ada user default di database");
            return ApiResponse<OrderDto>.FailResult("Sistem kasir sedang tidak siap menerima pesanan");
        }

        Table? table = null;
        if (dto.OrderType.Equals("DineIn", StringComparison.OrdinalIgnoreCase))
        {
            if (dto.TableId.HasValue && dto.TableId.Value > 0)
            {
                table = await _tableRepository.GetByIdAsync(dto.TableId.Value);
            }
            if (table == null && dto.TableNumber.HasValue && dto.TableNumber.Value > 0)
            {
                var tables = await _tableRepository.FindAsync(t => t.Number == dto.TableNumber.Value);
                table = tables.FirstOrDefault();
            }
            if (table == null && dto.TableId.HasValue && dto.TableId.Value > 0)
            {
                var tables = await _tableRepository.FindAsync(t => t.Number == dto.TableId.Value);
                table = tables.FirstOrDefault();
            }
            if (table == null && dto.TableNumber.HasValue && dto.TableNumber.Value > 0)
            {
                table = await _tableRepository.GetByIdAsync(dto.TableNumber.Value);
            }

            if (table == null)
            {
                // Ambil meja pertama jika tidak ditemukan nomor meja spesifik
                var allTables = await _tableRepository.GetAllAsync();
                table = allTables.FirstOrDefault();
            }

            if (table != null)
            {
                dto.TableId = table.Id;
                table.Status = "Occupied";
                table.UpdatedAt = DateTime.UtcNow;
                await _tableRepository.UpdateAsync(table);
                _logger.LogInformation("[ORDER] 🪑 Status Meja #{TableNumber} diubah menjadi Occupied oleh Guest Self-Order", table.Number);
            }
        }

        var menuIds = normalizedItems.Select(i => i.EffectiveMenuId).Distinct().ToList();
        var menus = (await _menuRepository.FindAsync(m => menuIds.Contains(m.Id))).ToDictionary(m => m.Id);

        foreach (var item in normalizedItems)
        {
            if (!menus.TryGetValue(item.EffectiveMenuId, out var menu))
            {
                return ApiResponse<OrderDto>.FailResult($"Menu dengan ID {item.EffectiveMenuId} tidak ditemukan");
            }

            if (!menu.IsAvailable)
            {
                return ApiResponse<OrderDto>.FailResult($"Menu '{menu.Name}' saat ini sedang tidak tersedia");
            }
        }

        var orderNumber = await _orderRepository.GenerateOrderNumberAsync();
        var customerName = !string.IsNullOrWhiteSpace(dto.CustomerName)
            ? dto.CustomerName.Trim()
            : (table != null ? $"Tamu Meja #{table.Number}" : "Tamu");

        var order = new Order
        {
            OrderNumber = orderNumber,
            UserId = defaultUser.Id,
            CustomerName = customerName,
            TableId = dto.OrderType.Equals("DineIn", StringComparison.OrdinalIgnoreCase) ? (table?.Id ?? dto.TableId) : null,
            OrderType = dto.OrderType.Equals("TakeAway", StringComparison.OrdinalIgnoreCase) ? "TakeAway" : "DineIn",
            Status = "Pending",
            CreatedAt = DateTime.UtcNow,
            OrderItems = new List<OrderItem>()
        };

        decimal totalAmount = 0;
        foreach (var item in normalizedItems)
        {
            var menu = menus[item.EffectiveMenuId];
            var qty = item.EffectiveQuantity;
            var subtotal = menu.Price * qty;
            totalAmount += subtotal;

            order.OrderItems.Add(new OrderItem
            {
                MenuId = menu.Id,
                Quantity = qty,
                Price = menu.Price,
                Subtotal = subtotal
            });
        }

        order.TotalAmount = totalAmount;

        var createdOrder = await _orderRepository.AddAsync(order);
        var fullOrder = await _orderRepository.GetOrderWithDetailsAsync(createdOrder.Id);

        _logger.LogInformation("[ORDER] ✅ Pesanan Self-Order berhasil dibuat! No: {OrderNumber}, Meja: #{Table}, Total: Rp {Total:N0}, Tamu: '{Customer}'",
            order.OrderNumber, table?.Number, order.TotalAmount, customerName);

        var orderDto = MapToOrderDto(fullOrder!);
        return ApiResponse<OrderDto>.SuccessResult(orderDto, "Pesanan berhasil dikirim ke Barista & Dapur");
    }

    public async Task<ApiResponse<OrderDto>> UpdateStatusAsync(int id, UpdateOrderStatusDto dto)
    {
        _logger.LogInformation("[ORDER] 🔄 Memperbarui status pesanan ID {OrderId} ke '{NewStatus}'", id, dto.Status);

        var order = await _orderRepository.GetOrderWithDetailsAsync(id);
        if (order == null)
        {
            _logger.LogWarning("[ORDER] ⚠️ Gagal update status: Pesanan ID {OrderId} tidak ditemukan", id);
            return ApiResponse<OrderDto>.FailResult("Pesanan tidak ditemukan");
        }

        var validStatuses = new[] { "Pending", "Preparing", "Ready", "Completed", "Cancelled" };
        if (!validStatuses.Contains(dto.Status, StringComparer.OrdinalIgnoreCase))
        {
            _logger.LogWarning("[ORDER] ⚠️ Gagal update status: Status '{InvalidStatus}' tidak valid", dto.Status);
            return ApiResponse<OrderDto>.FailResult("Status pesanan tidak valid");
        }

        var oldStatus = order.Status;
        order.Status = dto.Status;
        order.UpdatedAt = DateTime.UtcNow;

        // If completed or cancelled and has table, check if other active orders exist on that table
        if ((dto.Status.Equals("Completed", StringComparison.OrdinalIgnoreCase) || 
             dto.Status.Equals("Cancelled", StringComparison.OrdinalIgnoreCase)) && 
             order.TableId.HasValue)
        {
            var otherActiveOrders = await _orderRepository.Query()
                .AnyAsync(o => o.Id != id && o.TableId == order.TableId && (o.Status == "Pending" || o.Status == "Preparing" || o.Status == "Ready"));

            if (!otherActiveOrders)
            {
                var table = await _tableRepository.GetByIdAsync(order.TableId.Value);
                if (table != null && table.Status == "Occupied")
                {
                    table.Status = "Available";
                    table.UpdatedAt = DateTime.UtcNow;
                    await _tableRepository.UpdateAsync(table);
                    _logger.LogInformation("[ORDER] 🪑 Meja #{TableNumber} kembali menjadi Available (Pesanan selesai/dibatalkan)", table.Number);
                }
            }
        }

        await _orderRepository.UpdateAsync(order);
        var updated = await _orderRepository.GetOrderWithDetailsAsync(id);

        _logger.LogInformation("[ORDER] ✅ Status pesanan {OrderNumber} (ID: {OrderId}) berhasil diubah: '{OldStatus}' -> '{NewStatus}'",
            order.OrderNumber, order.Id, oldStatus, order.Status);

        return ApiResponse<OrderDto>.SuccessResult(MapToOrderDto(updated!), "Status pesanan berhasil diperbarui");
    }

    public async Task<ApiResponse<bool>> CancelOrderAsync(int id)
    {
        _logger.LogInformation("[ORDER] 🚫 Memproses pembatalan pesanan ID: {OrderId}", id);

        var order = await _orderRepository.GetOrderWithDetailsAsync(id);
        if (order == null)
        {
            _logger.LogWarning("[ORDER] ⚠️ Pembatalan gagal: Pesanan ID {OrderId} tidak ditemukan", id);
            return ApiResponse<bool>.FailResult("Pesanan tidak ditemukan");
        }

        if (order.Status == "Completed")
        {
            _logger.LogWarning("[ORDER] ⚠️ Pembatalan ditolak: Pesanan {OrderNumber} sudah Completed", order.OrderNumber);
            return ApiResponse<bool>.FailResult("Pesanan yang sudah selesai (Completed) tidak dapat dibatalkan");
        }

        order.Status = "Cancelled";
        order.UpdatedAt = DateTime.UtcNow;

        if (order.TableId.HasValue)
        {
            var otherActiveOrders = await _orderRepository.Query()
                .AnyAsync(o => o.Id != id && o.TableId == order.TableId && (o.Status == "Pending" || o.Status == "Preparing" || o.Status == "Ready"));

            if (!otherActiveOrders)
            {
                var table = await _tableRepository.GetByIdAsync(order.TableId.Value);
                if (table != null && table.Status == "Occupied")
                {
                    table.Status = "Available";
                    table.UpdatedAt = DateTime.UtcNow;
                    await _tableRepository.UpdateAsync(table);
                    _logger.LogInformation("[ORDER] 🪑 Meja #{TableNumber} kembali Available setelah pesanan {OrderNumber} dibatalkan", table.Number, order.OrderNumber);
                }
            }
        }

        await _orderRepository.UpdateAsync(order);
        _logger.LogInformation("[ORDER] ✅ Pesanan {OrderNumber} (ID: {OrderId}) berhasil dibatalkan", order.OrderNumber, order.Id);

        return ApiResponse<bool>.SuccessResult(true, "Pesanan berhasil dibatalkan");
    }

    private static OrderDto MapToOrderDto(Order order)
    {
        var customer = !string.IsNullOrWhiteSpace(order.CustomerName)
            ? order.CustomerName
            : (order.Table != null ? $"Tamu Meja #{order.Table.Number}" : (order.User?.Name ?? "Tamu"));

        return new OrderDto
        {
            Id = order.Id,
            OrderNumber = order.OrderNumber,
            UserId = order.UserId,
            UserName = order.User?.Name ?? string.Empty,
            CustomerName = customer,
            TableId = order.TableId,
            TableNumber = order.Table?.Number,
            OrderType = order.OrderType,
            Status = order.Status,
            TotalAmount = order.TotalAmount,
            CreatedAt = order.CreatedAt,
            UpdatedAt = order.UpdatedAt,
            Items = order.OrderItems.Select(oi => new OrderItemDto
            {
                Id = oi.Id,
                MenuId = oi.MenuId,
                MenuName = oi.Menu?.Name ?? string.Empty,
                ImageUrl = oi.Menu?.ImageUrl,
                Quantity = oi.Quantity,
                Price = oi.Price,
                Subtotal = oi.Subtotal,
                Menu = oi.Menu != null ? new DTOs.Menu.MenuDto
                {
                    Id = oi.Menu.Id,
                    CategoryId = oi.Menu.CategoryId,
                    CategoryName = oi.Menu.Category?.Name ?? string.Empty,
                    Name = oi.Menu.Name,
                    Description = oi.Menu.Description,
                    Price = oi.Menu.Price,
                    ImageUrl = oi.Menu.ImageUrl,
                    IsAvailable = oi.Menu.IsAvailable,
                    CreatedAt = oi.Menu.CreatedAt,
                    UpdatedAt = oi.Menu.UpdatedAt
                } : null
            }).ToList()
        };
    }
}

File: Uas/Caffèra — Cafe Management System/Frontend/src/services/orderService.js
import api, { getStorageData, setStorageData, STORAGE_KEYS } from './api';

export const orderService = {
  /**
   * Get all orders with filtering, searching, and pagination
   */
  getAll: async (params = {}) => {
    try {
      const response = await api.get('/orders', { params });
      if (response.data && response.data.success) {
        return response.data;
      }
    } catch (err) {
      console.warn('API /orders unavailable, using Mock Engine:', err.message);
    }

    let orders = getStorageData(STORAGE_KEYS.ORDERS);

    // Search filter
    if (params.search && params.search.trim()) {
      const query = params.search.toLowerCase().trim();
      orders = orders.filter(
        (o) =>
          o.orderNumber.toLowerCase().includes(query) ||
          (o.userName && o.userName.toLowerCase().includes(query)) ||
          (o.tableNumber && String(o.tableNumber).includes(query))
      );
    }

    // Status filter
    if (params.status && params.status !== 'All' && params.status.trim()) {
      orders = orders.filter(
        (o) => o.status.toLowerCase() === params.status.toLowerCase().trim()
      );
    }

    // Order Type filter
    if (params.orderType && params.orderType !== 'All' && params.orderType.trim()) {
      orders = orders.filter(
        (o) => o.orderType.toLowerCase() === params.orderType.toLowerCase().trim()
      );
    }

    // Table filter
    if (params.tableId && Number(params.tableId) > 0) {
      orders = orders.filter((o) => o.tableId === Number(params.tableId));
    }

    // Date range filter
    if (params.startDate) {
      const start = new Date(params.startDate);
      orders = orders.filter((o) => new Date(o.createdAt) >= start);
    }
    if (params.endDate) {
      const end = new Date(params.endDate);
      end.setHours(23, 59, 59, 999);
      orders = orders.filter((o) => new Date(o.createdAt) <= end);
    }

    // Sort by latest createdAt by default
    orders.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));

    const totalCount = orders.length;
    const pageNumber = Number(params.pageNumber) || 1;
    const pageSize = Number(params.pageSize) || 10;
    const startIndex = (pageNumber - 1) * pageSize;
    const pagedItems = orders.slice(startIndex, startIndex + pageSize);

    return {
      success: true,
      message: 'Berhasil mengambil daftar pesanan',
      data: {
        items: pagedItems,
        pageNumber,
        pageSize,
        totalCount,
        totalPages: Math.ceil(totalCount / pageSize),
        hasPreviousPage: pageNumber > 1,
        hasNextPage: pageNumber < Math.ceil(totalCount / pageSize),
      },
    };
  },

  /**
   * Get order by ID
   */
  getById: async (id) => {
    try {
      const response = await api.get(`/orders/${id}`);
      if (response.data && response.data.success) {
        return response.data;
      }
    } catch (err) {
      console.warn(`API /orders/${id} unavailable, using Mock Engine:`, err.message);
    }

    const orders = getStorageData(STORAGE_KEYS.ORDERS);
    const order = orders.find((o) => o.id === Number(id));

    if (!order) {
      return { success: false, message: 'Pesanan tidak ditemukan' };
    }

    return {
      success: true,
      data: order,
    };
  },

  /**
   * Create New Order (POS checkout)
   */
  create: async (dto) => {
    try {
      const response = await api.post('/orders', dto);
      if (response.data && response.data.success) {
        return response.data;
      }
    } catch (err) {
      console.warn('API POST /orders unavailable, using Mock Engine:', err.message);
    }

    const orders = getStorageData(STORAGE_KEYS.ORDERS);
    const menus = getStorageData(STORAGE_KEYS.MENUS);
    const tables = getStorageData(STORAGE_KEYS.TABLES);
    const users = getStorageData(STORAGE_KEYS.USERS);

    const currentUser = JSON.parse(localStorage.getItem('caffera_user') || 'null') || users[0];
    const table = dto.tableId ? tables.find((t) => t.id === Number(dto.tableId)) : null;

    const newId = orders.length > 0 ? Math.max(...orders.map((o) => o.id)) + 1 : 1;
    const todayStr = new Date().toISOString().slice(0, 10).replace(/-/g, '');
    const orderNumber = `ORD-${todayStr}-${String(newId).padStart(3, '0')}`;

    // Process order items
    let calculatedTotal = 0;
    const orderItems = (dto.items || []).map((item, idx) => {
      const menu = menus.find((m) => m.id === Number(item.menuId));
      const price = menu ? menu.price : item.price || 0;
      const quantity = Number(item.quantity) || 1;
      const subtotal = price * quantity;
      calculatedTotal += subtotal;

      return {
        id: idx + 1,
        orderId: newId,
        menuId: Number(item.menuId),
        menuName: menu ? menu.name : 'Unknown Item',
        quantity,
        price,
        subtotal,
      };
    });

    const newOrder = {
      id: newId,
      orderNumber,
      userId: currentUser.id,
      userName: currentUser.name,
      tableId: dto.tableId ? Number(dto.tableId) : null,
      tableNumber: table ? table.number : null,
      orderType: dto.orderType || 'DineIn',
      status: 'Pending',
      totalAmount: calculatedTotal,
      createdAt: new Date().toISOString(),
      items: orderItems,
    };

    // Update table status if DineIn
    if (table) {
      table.status = 'Occupied';
      setStorageData(STORAGE_KEYS.TABLES, tables);
    }

    orders.unshift(newOrder);
    setStorageData(STORAGE_KEYS.ORDERS, orders);

    return {
      success: true,
      message: 'Pesanan berhasil dibuat',
      data: newOrder,
    };
  },

  /**
   * Update Order Status
   */
  updateStatus: async (id, status) => {
    try {
      const response = await api.patch(`/orders/${id}/status`, { status });
      if (response.data && response.data.success) {
        return response.data;
      }
    } catch (err) {
      console.warn(`API PATCH /orders/${id}/status unavailable, using Mock Engine:`, err.message);
    }

    const orders = getStorageData(STORAGE_KEYS.ORDERS);
    const index = orders.findIndex((o) => o.id === Number(id));

    if (index === -1) {
      return { success: false, message: 'Pesanan tidak ditemukan' };
    }

    orders[index].status = status;
    orders[index].updatedAt = new Date().toISOString();

    // If order is completed or cancelled, optionally release table if no other active orders on that table
    if (status === 'Completed' || status === 'Cancelled') {
      const tableId = orders[index].tableId;
      if (tableId) {
        const tables = getStorageData(STORAGE_KEYS.TABLES);
        const tableIndex = tables.findIndex((t) => t.id === tableId);
        if (tableIndex !== -1) {
          const hasOtherActive = orders.some(
            (o) =>
              o.id !== Number(id) &&
              o.tableId === tableId &&
              (o.status === 'Pending' || o.status === 'Preparing' || o.status === 'Ready')
          );
          if (!hasOtherActive) {
            tables[tableIndex].status = 'Available';
            setStorageData(STORAGE_KEYS.TABLES, tables);
          }
        }
      }
    }

    setStorageData(STORAGE_KEYS.ORDERS, orders);

    return {
      success: true,
      message: `Status pesanan diubah menjadi ${status}`,
      data: orders[index],
    };
  },

  /**
   * Cancel / Delete Order
   */
  cancel: async (id) => {
    try {
      const response = await api.delete(`/orders/${id}/cancel`);
      if (response.data && response.data.success) {
        return response.data;
      }
    } catch (err) {
      console.warn(`API DELETE /orders/${id}/cancel unavailable, using Mock Engine:`, err.message);
    }

    return orderService.updateStatus(id, 'Cancelled');
  },
};
